crates/bevy_asset/src/io/wasm.rs

```diff
@@ -121,3 +123,115 @@ impl AssetReader for HttpWasmAssetReader {
         Ok(false)
     }
 }
+
+/// An [`AsyncRead`] implementation capable of reading a [`Uint8Array`].
+pub struct Uint8ArrayReader {
+    array: Uint8Array,
+    initial_offset: u32,
+}
+
+impl Uint8ArrayReader {
+    /// Create a new [`Uint8ArrayReader`] for `bytes`.
+    pub fn new(array: Uint8Array) -> Self {
+        Self {
+            initial_offset: array.byte_offset(),
+            array,
+        }
+    }
+}
+
+impl AsyncRead for Uint8ArrayReader {
+    fn poll_read(
+        mut self: Pin<&mut Self>,
+        _cx: &mut Context,
+        buf: &mut [u8],
+    ) -> Poll<futures_io::Result<usize>> {
+        let array_len = self.array.length();
+        let n = u32::min(buf.len() as u32, array_len);
+        self.array.subarray(0, n).copy_to(&mut buf[..n as usize]); // NOTE: copy_to will panic if the lengths do not exactly match
+        self.array = self.array.subarray(n, array_len);
+        Poll::Ready(Ok(n as usize))
+    }
+}
+
+impl AsyncSeek for Uint8ArrayReader {
+    fn poll_seek(
+        mut self: Pin<&mut Self>,
+        _cx: &mut Context,
+        seek_from: SeekFrom,
+    ) -> Poll<std::io::Result<u64>> {
+        let array_len = self.array.length();
+        let current_array_buffer_offset = self.array.byte_offset();
+        let array_buffer_end = current_array_buffer_offset + array_len;
+        let new_array_buffer_offset = match seek_from {
+            SeekFrom::Start(from_start) => self
+                .initial_offset
+                .saturating_add(u32::try_from(from_start).unwrap_or(u32::MAX))
+                .min(array_buffer_end),
+            SeekFrom::End(from_end) => {
+                if from_end.is_negative() {
+                    array_buffer_end
+                        .saturating_sub(u32::try_from(from_end.abs()).unwrap_or(u32::MAX))
+                        .max(self.initial_offset)
+                } else {
+                    array_buffer_end
+                }
+            }
+            SeekFrom::Current(from_current) => {
+                if from_current.is_negative() {
+                    current_array_buffer_offset
+                        .saturating_sub(u32::try_from(from_current.abs()).unwrap_or(u32::MAX))
+                        .max(self.initial_offset)
+                } else {
+                    current_array_buffer_offset
+                        .saturating_add(u32::try_from(from_current).unwrap_or(u32::MAX))
+                        .min(array_buffer_end)
+                }
+            }
+        };
+        debug_assert!(new_array_buffer_offset >= self.initial_offset);
+        debug_assert!(new_array_buffer_offset <= array_buffer_end);
+        self.array = self
+            .array
+            .constructor()
+            .call3(
+                &JsValue::UNDEFINED,
+                &self.array.buffer(),
+                &new_array_buffer_offset.into(),
+                &array_buffer_end.into(),
+            )
```

Comment:
I'm not exactly a Wasm expert, could this be `Uint8Array::new_with_byte_offset(self.array.buffer(), 0)`? How does this differ?

Reply:
The difference is I didn't see that method when scanning the docs 🤣 . Will change to that